source/Hyperspace-Model/Retry.class.st

```diff
@@ -0,0 +1,178 @@
+```
+
+
+"
+Class {
+	#name : #Retry,
+	#superclass : #Object,
+	#instVars : [
+		'maximumNumberOfAttempts',
+		'waitActionBetweenRetries',
+		'exceptionsToIgnore',
+		'retryCondition',
+		'conditionBasedActionBetweenRetries',
+		'exceptionBasedActionBetweenRetries',
+		'exceptionToRetry'
+	],
+	#category : #'Hyperspace-Model-Resilience'
+}
+
+{ #category : #'instance creation' }
+Retry class >> configuredBy: configurationAction [
+
+	^ self new initializeConfiguredBy: configurationAction
+]
+
+{ #category : #accessing }
+Retry class >> defaultMaximumNumberOfAttempts [
+
+	^ 3
+]
+
+{ #category : #evaluating }
+Retry class >> value: aNyladicClosure [
+
+	^ self value: aNyladicClosure configuredBy: [ :retry | retry onException ]
+]
+
+{ #category : #evaluating }
+Retry class >> value: aNiladycBlock configuredBy: configurationAction [
+
+	^ ( self configuredBy: configurationAction ) value: aNiladycBlock
+]
+
+{ #category : #configuring }
+Retry >> backoffExponentiallyWithTimeSlot: aDuration [
+
+	self backoffExponentiallyWithTimeSlot: aDuration using: [ :integer | integer atRandom ]
+]
+
+{ #category : #configuring }
+Retry >> backoffExponentiallyWithTimeSlot: aDuration using: aRandomizingAction [
+
+	"This strategy implements [Exponential Backoff](https://en.wikipedia.org/wiki/Exponential_backoff).
+	
+	`aRandomizingAction` must provide a randomization block that will be evaluated with an integer value 
+	and must answer a random integer between 1 and that value"
+
+	waitActionBetweenRetries := [ :currentAttemptCount | 
+	                            | timeSlotCount |
+	                            "We are applying the randomization function before subtracting one because #atRandom on an integer works in the range [1,self].
+										In other case you will never get a zero time slot"
+	                            timeSlotCount := ( aRandomizingAction value:
+		                                               ( 2 raisedTo: currentAttemptCount ) ) - 1.
+	                            ( aDuration * timeSlotCount ) wait
+	                            ]
+]
+
+{ #category : #configuring }
+Retry >> every: aDuration [
+
+	waitActionBetweenRetries := [ aDuration wait ]
+]
+
+{ #category : #configuring }
+Retry >> ignore: anException [
+
+	exceptionsToIgnore := exceptionsToIgnore , anException
+]
+
+{ #category : #initialization }
+Retry >> initializeConfiguredBy: configurationAction [
+
+	maximumNumberOfAttempts := self class defaultMaximumNumberOfAttempts.
+	waitActionBetweenRetries := [  ].
+	conditionBasedActionBetweenRetries := [  ].
+	exceptionBasedActionBetweenRetries := [  ].
+	exceptionToRetry := ExceptionSet new.
+	exceptionsToIgnore := ExceptionSet new.
+	self when: [ :result | false ].
+	configurationAction value: self
+]
+
+{ #category : #configuring }
+Retry >> on: anException [
+
+	exceptionToRetry := anException
+]
+
+{ #category : #configuring }
+Retry >> on: anException evaluating: aDyadicBlock [
```

Comment:
I was expecting being able to add multiple handlers. The same apply to #on: message. Now I get why it's possible to send #ignore: multiple times. 

Does it makes sense to have more than one handler by adding #on:ignoring: and #on:ignoring:evaluating: messages?

Reply:
At least for all the use cases we have now nobody is doing something specific for every exception. The block is not intended to change behavior but to produce logging or collect information about each retry made.